### ISSUE TITLE
If loaded blueprint has nonconsecutive `entity_numbers`, newly placed entities will overwrite existing ones

### ISSUE THREAD

**Issue by xiagu** [Teoxoy/factorio-blueprint-editor#5]
### TL;DR
If you ever delete entities (say, because you made a mistake) and then place more, then copying and pasting the generated string into the app and working on it further will make you overwrite entities.

### Repro
Here's a simple repro [(link to app with blueprint loaded)](https://teoxoy.github.io/factorio-blueprint-editor/?source=0eJyV0sGKwjAQBuB3+c8paNVdyXFfY5ElrYMOJNOSRNlS8u6mLSyyoq2nkEzyzSSZHpW9UOtZInQPrhsJ0N89Ap/E2GEtdi1BgyM5KIhxwyx6I6FtfCwqshFJgeVIv9DrpGYPmxDIVZblVDhTn1moKO+IMh0USCJHpqmYcdL9yMVV5HOOZ2UotE3IxxoZcmeq2Ch0edhm/cie6im2Gor8h+6Wo+sRXc2bH4vNxeT23bsvMD9f/8yjXI7yZl7evyk/gXM/jB2k77pV4Uo+TBsUrMmPkONff/GUblwi9Wg=):
```
0eJyV0sGKwjAQBuB3+c8paNVdyXFfY5ElrYMOJNOSRNlS8u6mLSyyoq2nkEzyzSSZHpW9UOtZInQPrhsJ0N89Ap/E2GEtdi1BgyM5KIhxwyx6I6FtfCwqshFJgeVIv9DrpGYPmxDIVZblVDhTn1moKO+IMh0USCJHpqmYcdL9yMVV5HOOZ2UotE3IxxoZcmeq2Ch0edhm/cie6im2Gor8h+6Wo+sRXc2bH4vNxeT23bsvMD9f/8yjXI7yZl7evyk/gXM/jB2k77pV4Uo+TBsUrMmPkONff/GUblwi9Wg=
```
I placed 6 transport belts, deleted 2, then placed 2 assemblers. The assemblers have `entity_numbers` 7 (left) and 8 (right), but there are only 6 entities in the blueprint. Adding a 7th entity after loading this blueprint will overwrite the left assembler (you can verify by pipetting it). Deleting the new 7th entity (not the assembler) will remove entry 7 from the blueprint's `rawEntities` map, and cause errors whenever you try to do anything with the "ghost" assembler.

### Longer explanation
When a blueprint is pasted in, `next_entity_number` is set to the number of entities in the blueprint + 1. This works fine if all the `entity_numbers` in the blueprint are consecutive, with no gaps. But it's easy to end up with gaps, and `entity_numbers` larger than the number of entities.

The blueprint `next_entity_number` is monotonically increasing. Deleting entities removes them from the `rawEntities` map, but doesn't change `next_entity_number`. So any blueprint built in the app where the user deleted an entity and then added more will have this property (and run into this problem).

### Thoughts on fix
- When generating a blueprint string, "compress" the entity numbers to remove all gaps.
- In order for old blueprints made with this app (and any other source with nonconsecutive entity numbers) to work, when importing a blueprint string, set `next_entity_number` to the largest entity number + 1 instead of the length + 1.

**Comment by Teoxoy** [comment]
Thanks for reporting this!
I fixed it by regenerating the entity numbers at export. By taking this approach, I also had to assign the new entity number in the connections object so that nothing breaks.

**Comment by Teoxoy** [comment]
Some blueprints I encountered online have nonconsecutive `entity_numbers`s so I applied the same function at import time.
